src/SSHDebugPS/SSH/SSHPortSupplier.cs

```diff
@@ -95,6 +95,12 @@ private bool IsLibLinuxAvailable()
             return (ServiceProvider.GlobalProvider.GetService(typeof(IVsConnectionManager)) as IVsConnectionManager) != null;
         }
 
+        public override int CanPersistPorts()
+        {
+            return HR.S_OK;
+        }
+
```

Comment:
The SSH port supplier doesn't persist ports. This should be returning S_FALSE. So that the SDM will handle persistence.

Reply:
Pretty sure they (liblinux) persist the SSH connections so we by default do because we query them.

Reply:
Got it. In that case we could stop implementing IDebugPortSupplier3, and EnumPersistedPorts. EnumPersistedPorts will only ever be called if CanPersistPorts returns S_FALSE.

---
In reply to: [287537299](https://github.com/microsoft/MIEngine/pull/868#discussion_r287537299) [](ancestors = 287537299)

Reply:
I need IDebugPortSupplier3 for CanPersistPorts though. I thought EnumPersistedPorts is only called if CanPersistPorts is true?

---
In reply to: [287537461](https://github.com/microsoft/MIEngine/pull/868#discussion_r287537461) [](ancestors = 287537461,287537299)

Reply:
No, EnumPersistedPorts is only called if CanPersistPorts returns S_FALSE.

Reply:
I updated the comments in msdbg.idl to make this less confusing:

```C++
// --------------------------------------------------------
// IDebugPortSupplier3
//
// Optional interface which can be implemented by port suppliers to opt-into
// have the SDM persist a MRU of port names.
[
	object,
	uuid(5b5eec44-51aa-4210-b84f-1938b8576d8d),
	pointer_default(unique)
]
interface IDebugPortSupplier3: IDebugPortSupplier2
{
	// Can the port supplier remember new ports? S_OK == yes; S_FALSE == no
        //
        // Returning S_FALSE from this method indicates that the SDM should persist
        // port names for the port supplier, and that the SDM can call EnumPersistedPorts
        // with the names that it has persisted.
	HRESULT CanPersistPorts();
	
	// Convert the port names that the SDM has persisted into an enumerator of IDebugPort2
        // objects. This method will only be called if `CanPersistPorts` returns S_FALSE.
	HRESULT EnumPersistedPorts(
		[in] BSTR_ARRAY PortNames,
		[out] IEnumDebugPorts2** ppEnum);
};
```

Reply:
That makes more sense. I think in the Docker scenario we should return FALSE which will force a call to EnumPersistedPorts. If that isn't the behavior today, I'll fix it for GA